models/locked_resource.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/setting"
+)
+
+// LockedResource represents the locking key for a pessimistic
+// lock that can hold a counter
+type LockedResource struct {
+	LockType string `xorm:"pk VARCHAR(30)"`
+	LockKey  int64  `xorm:"pk"`
+	Counter  int64  `xorm:"NOT NULL DEFAULT 0"`
+
+	engine Engine `xorm:"-"`
+}
+
+// GetLockedResource gets or creates a pessimistic lock on the given type and key
+func GetLockedResource(e Engine, lockType string, lockKey int64) (*LockedResource, error) {
+	resource := &LockedResource{LockType: lockType, LockKey: lockKey}
+
+	if err := upsertLockedResource(e, resource); err != nil {
+		return nil, fmt.Errorf("upsertLockedResource: %v", err)
+	}
+
+	// Read back the record we've created or locked to get the current Counter value
+	if has, err := e.Table(resource).NoCache().NoAutoCondition().AllCols().
+		Where("lock_type = ? AND lock_key = ?", lockType, lockKey).Get(resource); err != nil {
+		return nil, fmt.Errorf("get locked resource %s:%d: %v", lockType, lockKey, err)
+	} else if !has {
+		return nil, fmt.Errorf("unexpected upsert fail  %s:%d", lockType, lockKey)
+	}
+
+	// Once active, the locked resource is tied to a specific session
+	resource.engine = e
+
+	return resource, nil
+}
+
+// UpdateValue updates the value of the counter of a locked resource
+func (r *LockedResource) UpdateValue() error {
```

Comment:
How did we lock the counter?

Reply:
> How did we lock the counter?

@lunny at the upsert in `GetLockedResource`. Being an insert/update, the record remains locked for the rest of the transaction. The counter is a column of the `locked_resource` table.

Reply:
The transaction in fact will not lock your record I think.

Reply:
> 
> 
> The transaction in fact will not lock your record I think.

@lunny I've written tests to prove that 😁. Check [`models/integrations/locked_resource_test.go`](https://github.com/guillep2k/gitea/blob/bylock-indexes/integrations/locked_resource_test.go).